File: ASIA PACIFIC/SOUTH PACIFIC/PROBLEM/Pokemon.py
def memo(pos, weight, budget, best, Cost, Prob): 
    if pos == 0 and weight == 0: return 1.0
    if pos == 0 : return 0.0
    ret = best[pos][weight]
    if ret >= 0: return ret 
    ret = 0
    acc = 1.0
    for k in range(budget + 1): 
        if weight >= k * Cost[pos] : 
            ret = max(ret, memo(pos - 1, weight - k * Cost[pos], budget, best, Cost, Prob) * (1.0 - acc))
        acc = acc * (1.0 - Prob[pos])
    best[pos][weight] = ret
    return ret

def solve(): 
    budget, n = map(int, input().split())
    if budget == 0 : exit(0)
    Prob = [0.0 for i in range(n + 1)]
    Cost = [0 for i in range(n + 1)]
    for i in range(1, n + 1): 
        val, p = map(str, input().split()) 
        Cost[i] = int(val)
        Prob[i] = float(p)
    best = [[-1.0 for i in range(budget + 1)] for j in range(n + 1)] 
    def memo(pos, weight): 
        if pos == 0 and weight == 0: return 1.0
        if pos == 0 : return 0.0
        ret = best[pos][weight]
        if ret >= 0: return ret 
        ret = 0
        acc = 1.0
        for k in range(budget + 1): 
            if weight >= k * Cost[pos] : 
                ret = max(ret, memo(pos - 1, weight - k * Cost[pos]) * (1.0 - acc))
            else: 
                break
            acc = acc * (1.0 - Prob[pos])
        best[pos][weight] = ret
        return ret
    ret = 0.0 
    ind = 0
    for w in range(budget + 1): 
        opt = memo(n, w)
        if ret < opt: 
           ret, ind = opt, w
    print(ind, "{:.4f}".format(ret))

while True : 
    solve()

File: ASIA PACIFIC/SOUTH PACIFIC/PROBLEM/Pokemon.cc
#include "bits/stdc++.h"

using namespace std; 

const int maxf = 10010; 

int N, B; 
double dyn[42][maxf]; 
double Prob[42];
int cost[42]; 

double memo(int pos, int budget) {
  if(pos == 0) return (budget == 0) * 1.0; 
  double &ret = dyn[pos][budget];   
  if(ret >= 0) return ret;
  ret = 0; 
  double acc = 1; 
  for(int k = 0; k * cost[pos] <= budget; k++) { 
    ret = max(ret, memo(pos - 1, budget - k * cost[pos]) * (1.0 - acc)); 
    acc = acc * (1.0 - Prob[pos]); 
  }
  return ret; 
}

int main(void) {
  ios_base::sync_with_stdio(false); 
  cin.tie(0); 
  cout.tie(0); 
  while(cin >> B >> N && N) { 
    for(int i = 1; i <= N; i++) { 
      cin >> cost[i] >> Prob[i]; 
    }
    for(int i = 0; i < maxf; i++) { 
      for(int j = 0; j < 32; j++) { 
        dyn[j][i] = -1.0; 
      }
    }
    double ret = 0; 
    int pos; 
    for(int w = 0; w <= B; w++) { 
      if(ret < memo(N, w)) { 
        ret = memo(N, w); 
        pos = w; 
      }
    }
    cout << pos << " " << fixed << setprecision(4) << ret << '\n'; 
  } 
  return 0; 
}

